File: docs/algorithms/code/algorithms/countNodes.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function countNodes(root: TreeNode | null): number {
  if (!root) {
    return 0;
  }

  const stack: TreeNode[] = [root];

  let count = 0;

  while (stack.length) {
    const size = stack.length;

    for (let i = 0; i < size; i++) {
      const node = stack.shift();
      count++;

      if (node.left) {
        stack.push(node.left);
      }

      if (node.right) {
        stack.push(node.right);
      }
    }
  }

  return count;
}

function countNodes(root: TreeNode | null): number {
  return !root ? 0 : countNodes(root.left) + countNodes(root.right) + 1;
}

File: docs/algorithms/code/algorithms/countNodes.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func countNodes(root *TreeNode) int {
	if root == nil {
		return 0
	}

	leftDepth, rightDepth := 0, 0
	leftNode, rightNode := root.Left, root.Right
	for leftNode != nil {
		leftNode = leftNode.Left
		leftDepth++
	}
	for rightNode != nil {
		rightNode = rightNode.Right
		rightDepth++
	}
	if leftDepth == rightDepth {
		return 2 << leftDepth - 1
	}
	return countNodes(root.Left) + countNodes(root.Right) + 1
}